File: MasteringDart/lib/chapters/chapter08/intermediate/intermediate_classes.dart
// ignore_for_file: avoid_print
//
//  intermediate_classes.dart
//  MasteringDart
//
//  Capítulo 08 — Clases y objetos.
//  Nivel INTERMEDIO: Clases con métodos (Product, Car).
//
//  Ejecutar con: dart run lib/chapters/chapter08/intermediate/intermediate_classes.dart

/// Clase Product con name y price. Método para calcular precio con descuento.
class Product {
  final String name;
  final double price;

  Product(this.name, this.price);

  /// Devuelve el precio después de aplicar un porcentaje de descuento (ej. 0.2 = 20% de descuento).
  double priceAfterDiscount(double discountPercent) {
    return price * (1 - discountPercent);
  }
}

/// Clase Car con brand y model. Tiene método start().
class Car {
  final String brand;
  final String model;

  Car(this.brand, this.model);

  void start() {
    print('$brand $model is starting...');
  }
}

/// Ejecuta ejemplos intermedios de clases.
void runIntermediateClassExamples() {
  final product = Product('Laptop', 999.99);
  print('Product: ${product.name}, price: \$${product.price}');
  final discounted = product.priceAfterDiscount(0.15); // 15% de descuento
  print('Price after 15% discount: \$$discounted');

  final car = Car('Toyota', 'Camry');
  print('Car: ${car.brand} ${car.model}');
  car.start();
}

void main() {
  runIntermediateClassExamples();
}

File: MasteringKotlin/app/src/main/java/com/devlokos/masteringkotlin/chapters/chapter08/intermediate/IntermediateClasses.kt
package com.devlokos.masteringkotlin.chapters.chapter08.intermediate

/**
 * Capítulo 08 — Clases y Objetos.
 * Nivel INTERMEDIO: Clases con métodos y lógica simple.
 *
 * Ejemplos: Product (name, price, discount), Car (brand, model, start).
 */
fun main() {
    runIntermediateClassExamples()
}

/**
 * Clase Product con name y price. Tiene un método para calcular el precio con descuento.
 */
class Product(val name: String, val price: Double) {
    /** Devuelve el precio tras aplicar un porcentaje de descuento (ej. 0.2 = 20% off). */
    fun priceAfterDiscount(discountPercent: Double): Double {
        return price * (1 - discountPercent)
    }
}

/**
 * Clase Car con brand y model. Tiene el método start().
 */
class Car(val brand: String, val model: String) {
    fun start() {
        println("$brand $model is starting...")
    }
}

fun runIntermediateClassExamples() {
    // --- Product: propiedades y método ---
    val product = Product("Laptop", 999.99)
    println("Product: ${product.name}, price: \$${product.price}")
    val discounted = product.priceAfterDiscount(0.15) // 15% off
    println("Price after 15% discount: \$$discounted")

    // --- Car: propiedades y método start() ---
    val car = Car("Toyota", "Camry")
    println("Car: ${car.brand} ${car.model}")
    car.start()
}
